common/rpc/interceptor/namespace_handover.go

```diff
@@ -0,0 +1,144 @@
+package interceptor
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/api"
+	"go.temporal.io/server/common/clock"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	"google.golang.org/grpc"
+)
+
+const (
+	ctxTailRoom = time.Millisecond * 100
+)
+
+var _ grpc.UnaryServerInterceptor = (*NamespaceHandoverInterceptor)(nil).Intercept
+
+type (
+	// NamespaceHandoverInterceptor handles the namespace in handover replication state
+	NamespaceHandoverInterceptor struct {
+		namespaceRegistry namespace.Registry
+		timeSource        clock.TimeSource
+		enabledForNS      dynamicconfig.BoolPropertyFnWithNamespaceFilter
+		metricsHandler    metrics.Handler
+		logger            log.Logger
+	}
+)
+
+func NewNamespaceHandoverInterceptor(
+	enabledForNS dynamicconfig.BoolPropertyFnWithNamespaceFilter,
+	namespaceRegistry namespace.Registry,
+	metricsHandler metrics.Handler,
+	logger log.Logger,
+	timeSource clock.TimeSource,
+) *NamespaceHandoverInterceptor {
+
+	return &NamespaceHandoverInterceptor{
+		enabledForNS:      enabledForNS,
+		namespaceRegistry: namespaceRegistry,
+		metricsHandler:    metricsHandler,
+		logger:            logger,
+		timeSource:        timeSource,
+	}
+}
+
+func (i *NamespaceHandoverInterceptor) Intercept(
+	ctx context.Context,
+	req any,
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (_ any, retError error) {
+	defer log.CapturePanic(i.logger, &retError)
+
+	if !strings.HasPrefix(info.FullMethod, api.WorkflowServicePrefix) {
+		return handler(ctx, req)
+	}
+
+	// review which method is allowed
+	methodName := api.MethodName(info.FullMethod)
+	namespaceName, err := GetNamespaceName(i.namespaceRegistry, req)
+	if err != nil {
+		return nil, err
+	}
+
+	if i.enabledForNS(namespaceName.String()) {
+		startTime := i.timeSource.Now()
+		defer func() {
+			metrics.HandoverWaitLatency.With(i.metricsHandler).Record(time.Since(startTime))
```

Comment:
I guess we only want to emit the metric when namespace actually waited on the handover state.